OpenSim/Common/Component.cpp

```diff
@@ -319,6 +331,11 @@ void Component::disconnect()
 
 void Component::addToSystem(SimTK::MultibodySystem& system) const
 {
+    // If being asked to be added to the same System that it is already
+    // a part, there is nothing to be done.
+    if (hasSystem() && (&getSystem() == &system)) {
+        return;
+    }
```

Comment:
Now that Component writers can dictate the order of subcomponents this will often mean cycling through Component lists of specific types, and it will be difficult to to guarantee they are attempting to add components only once. Since property changes should wipe out the system, this should be safe to do. My inclination is that it is too harsh to throw an Exception in this case.

`Component::setNextSubcomponentInSystem()` only appends the component to the list of ordered subcomponents if the subcomponent is not already in the list, so it is unique for a given Component. However, each subcomponent also has its own list.

But there is nothing stopping the component from directly calling `addToSystem` on a subcomponent. Note that previously, we didn't check and simply added the component to the system again. So this at least fixes that problem.

Here is the use case. A Model has Ground, Bodies, PhysicalOffsetFrames (POFs), Joints, Forces and user defined Components (like a Device). Some Joints have their own POFs (as subcomponents) and others connect to POFs somewhere else in the Model (like on Bodies or a Device). The MultibodyGraphMaker dictates which Bodies and POFs are to be added first to define each Joint in multibody tree order until the POFs, Bodies and Joints that make up the multibody tree are part of the System. At this point there are still POFs (not specifically part of the multibody tree) that have not been added to the System yet (think POF of the Device for attaching it). So a reasonable thing to do is to cycle through all POFs and add them next sinces, POFs act as attachment sites for other components (e.g. Stations, Forces, Device). POFs that were added as a requirement for the building the tree will already be present and I think it makes sense to now ignore them as the Model does another pass to make sure all POFs are added before attempting to add other Components to the System.

Let me know if you agree.